Use `RateLimitService` for all rate-limiting

robknight opened the pull request:
Closes #1379 

We currently have two separate rate-limiting mechanisms, one for account resets and a generic rate limiting service which is used for email token generation and verification. Account resets are limited to 5 per day, whereas token generation and verification are limited to 10 per hour. Account resets are tracked via a field in the `users` table, whereas token generation/verification is tracked in the `rate_limit_buckets` table.

This PR makes the `RateLimitService` more flexible, allowing the specification of a `maxActions` and `timePeriod` option for each rate limit type. This allows the `RateLimitService` to take over the job performed by the custom code managing rate-limiting for account resets. The call to the `RateLimitService` is a drop-in replacement to the check being done during account reset, and allows for all of the account-reset-specific rate limiting code to be removed.

**UPDATE**: the entire rate limiting logic is now implemented in TypeScript. The DB is only used for persistence of rate limit state.

**FURTHER UPDATE**: the rate limiting logic is back in SQL again, though this time just an `INSERT...ON CONFLICT DO UPDATE`.

artwyman commented on apps/passport-server/migrations/61_rate_limit_variable_periods.sql:
```diff
@@ -0,0 +1,63 @@
+    this_refill TIMESTAMPTZ;
+BEGIN
+    -- Lock the buckets until end of transaction
+    LOCK TABLE rate_limit_buckets IN EXCLUSIVE MODE;
```
Locking the whole table, as opposed to a single row, seems like it could cause contention problems if we use rate limiting for something high-rate, like request/transaction limits.  Would a SELECT FOR UPDATE below be sufficient?

artwyman commented on apps/passport-server/migrations/61_rate_limit_variable_periods.sql:
```diff
@@ -0,0 +1,63 @@
+-- Moving rate limit configuration to TypeScript code, so this is no longer
```
Now that we're re-implementing this rather than borrowing an off-the-shelf solution, I wonder if there's a strong need for doing this in a SQL function rather than Typescript.  I can see a performance argument (fewer round-trips to the database), but it's offset by my worry that I don't know enough about SQL functions to review this properly.  While we do have some decent unit tests, I don't trust them to be able to fully cover all the edge-cases and race-conditions at this level.

artwyman replied:
Clarifying after discussion on Discord: This comment is about the SQL function not the SQL table.  It's the logic in SQL which makes me nervous.  The state needs to be in SQL to persist against restarts and (eventually) be shared between multiple server instances.

This isn't a strong suggestion, just expressing that I'm nervous, since I'm not familiar enough with the tradeoffs and risks of SQL functions to be sure what's safe.  That said I realize that some tricky complexity around SQL is needed regardless, e.g. in the form of locks to ensure the logic is safely atomic.

artwyman commented on apps/passport-server/migrations/61_rate_limit_variable_periods.sql:
```diff
@@ -0,0 +1,63 @@
+  action_type VARCHAR, -- The type of action, such as "ACCOUNT_RESET"
+  action_id VARCHAR, -- A unique identifier, such as the account ID being reset
+  periodic_limit INTEGER, -- How many of these actions are permitted per period
+  time_period_seconds INTEGER, -- The time period, in seconds
```
Why seconds vs. ms?  It's probably short enough, but I could imagine use cases of a token bucket where sub-second accuracy is important.

robknight commented:
I've re-written this in TypeScript, and removed the SQL procedure.

Rate limits are tracked in-memory and in the DB. After each request, if the state of the rate limit bucket has changed, the rate limit service persists the new state. On startup, it loads the saved state when initializing the rate limiters. Expired state gets deleted once per hour.

This results in the same number of round-trips to the DB as previously, and even lets us skip hitting the DB at all in cases where the rate limit has already been exceeded (if the persisted record already says "remaining = 0" then there's nothing to update).

artwyman reviewed (changes requested):
I think this the latest revision is a more radical change to the functionality than I was expecting, and not in a direction I agree with.  To confirm my understanding, I think there are two major changes here:

1) The database is no longer the primary "live" state of rate limiting.  The live state is in memory, and backed up to the database after changes.  The database is never read after startup time.  This makes rate limiting explicitly not robust to multiple server instances (machines or processes) which seems like a step in the wrong direction.  Even if we have only one instance now, I think multiple instances is somewhere we should anticipate in future.

2) Rather than a token bucket with an even refill rate over time, this implementation seems to be bursty.  I.e. the bucket isn't refilled at all until the time expires, then a full allocation of tokens appears all at once.  This isn't an unacceptable solution for the current needs, but it seems worse and less flexible in general.

What I had in mind for a shift to TypeScript implementation was that the typescript function would be performing the same database updates as the SQL function, with appropriate locking, so that multiple servers could safely use the same database.  My discomfort with the complex SQL implementation was more of a nicety - I think I'd accept the SQL function if the calculations were put through some more careful review with docs.  Do you think an alternative approach is better?

robknight commented:
> I think this the latest revision is a more radical change to the functionality than I was expecting, and not in a direction I agree with. To confirm my understanding, I think there are two major changes here:
> 
> 1. The database is no longer the primary "live" state of rate limiting.  The live state is in memory, and backed up to the database after changes.  The database is never read after startup time.  This makes rate limiting explicitly not robust to multiple server instances (machines or processes) which seems like a step in the wrong direction.  Even if we have only one instance now, I think multiple instances is somewhere we should anticipate in future.
> 

The challenge here is that we want to do the following:
1) Establish the current state of the rate limiter
2) Perform some computation to work out, based on that current state, whether to allow or deny some action
3) Persist the new rate limiter state (if any)

And we also want to do this without any race conditions within or between these steps.

We have two places where there is natural safety from concurrency: our single-threaded NodeJS server (iff the code is not async code), and inside a Postgres transaction with a table lock. There are also two places where we can put the logic: in NodeJS, or in Postgres. To me it seemed desirable to do both of these in the same place, with the alternative being that one system performs the calculation while the other one ensures the atomicity.

And so my first version of this PR put both of these in Postgres, while NodeJS knows nothing about how the result is obtained, simply receiving a yes/no response. There are no race conditions because the Postgres server locks the table before updating it, and the lock is automatically released at the end of the transaction.

My second version simplified the Postgres part, and let NodeJS pass in some additional configuration variables, but still left Postgres handling the rate limiting logic, atomicity, and persistence. It simplified the stored procedure by removing the configuration-handling, and added more detailed comments to each step of the rate limiting logic.

Moving the logic to NodeJS means that we don't get the natural atomicity of running everything inside a Postgres transaction. But, since NodeJS is single-threaded, and synchronous code is therefore safe against race conditions, we can treat the rate limiter calcuations as equivalently safe. (Though as you point out, it is not safe against architectural change in the application, if we end up having multiple servers needing to share a single rate limit).

As I understand it, the other option would be to have NodeJS and Postgres coordinate on these two tasks. NodeJS should begin a transaction, then lock the table, then query the table, then perform the computation, then either `INSERT` or `UPDATE` the relevant data, then commit the transaction. That means five round-trips for every rate-limit check, rather than the original one.

Whereas the Postgres function table lock is automatically released without further communication, the lock taken out by NodeJS is not. We have to be very certain that our NodeJS server can't crash or otherwise misbehave while holding the lock. Since the database query functions it needs to call while holding the lock are async, we cannot prevent opportunities for some other piece of processing to take over, which could lead to the lock being held for longer than we would prefer. This is especially undesirable in rate-limiting code, where we might reasonably expect there to be an unusually high number of concurrent requests. With a bit of bad luck, we could also end up holding the lock while our single NodeJS thread gets stuck in an expensive PCD proving operation. I don't think there's any possibility of a deadlock here, but it seems likely to end up with more time spent waiting to acquire locks than is necessary, and since this lock acts on all rate-limited actions across all users, it's good to avoid unnecessary waits.

There's also the slightly spooky "Connection terminated unexpectedly" error that we still get on our database queries, which hasn't gone away. It has been hidden behind some [exponential back-off retry logic](https://github.com/proofcarryingdata/zupass/commit/efc22ec720882c3ac3b6351c05b5e7f04233d0e9) on all DB queries, and choosing not to log the error any more. I don't _know_ that this might strike in the middle of such a transaction and what the effect would be if it did, but I don't know that it can't either.

For these reasons, I preferred to put control decisively on one side or the other of the NodeJS/Postgres boundary, rather than having them coordinate a more complex process between each other. One side does something smart, and the other side does something simple, and there are only very simple points of coordination, with at most one interaction per rate limit check.

> 2. Rather than a token bucket with an even refill rate over time, this implementation seems to be bursty.  I.e. the bucket isn't refilled at all until the time expires, then a full allocation of tokens appears all at once.  This isn't an unacceptable solution for the current needs, but it seems worse and less flexible in general.

I did consider re-implementing the token bucket approach in TypeScript, but given that ease of comprehensibility was a concern with the earlier solution I thought it might be better to use something simpler. I am very much agnostic on this question as I don't think either approach has sufficiently great strengths or weaknesses to outweigh ease of understanding - I quite like the elegance of the token bucket as a matter of personal preference, though.

> What I had in mind for a shift to TypeScript implementation was that the typescript function would be performing the same database updates as the SQL function, with appropriate locking, so that multiple servers could safely use the same database. My discomfort with the complex SQL implementation was more of a nicety - I think I'd accept the SQL function if the calculations were put through some more careful review with docs. Do you think an alternative approach is better?

For the sake of completeness, now that we've moved token bucket configuration out of Postgres, there is also the option of doing whole thing with a single `INSERT... ON CONFLICT DO UPDATE` query. The only problem is that it's 24 lines long and even harder to read than the original SQL function, mostly because fitting it into a single command means we can't use variable names:
```sql
    INSERT INTO rate_limit_buckets (action_type, action_id, remaining, next_refill)
    VALUES ($1, $2, $3 - 1, TO_TIMESTAMP($5::DOUBLE PRECISION + $4::DOUBLE PRECISION))
    ON CONFLICT (action_type, action_id) DO
    UPDATE SET
    remaining = GREATEST(-1, 
      CASE WHEN TO_TIMESTAMP($5) < rate_limit_buckets.next_refill
      THEN
        rate_limit_buckets.remaining - 1
      ELSE
        LEAST(
          $3,
          rate_limit_buckets.remaining + 1
            + FLOOR($5 - EXTRACT(EPOCH FROM rate_limit_buckets.next_refill) / $4)::INTEGER)
        - 1
      END),
    next_refill =
    CASE 
      WHEN (TO_TIMESTAMP($5) >= rate_limit_buckets.next_refill)
        THEN TO_TIMESTAMP($5) + MAKE_INTERVAL(secs => EXTRACT(EPOCH FROM rate_limit_buckets.next_refill) % $4::INTEGER)
      WHEN (rate_limit_buckets.remaining > 0)
        THEN rate_limit_buckets.next_refill + MAKE_INTERVAL(secs => $4)
      ELSE rate_limit_buckets.next_refill
    END
    RETURNING *
```

Having considered all of the alternatives, on balance I think I prefer the version in the second commit. But then we have the question of what kind of documentation or testing would make this less fear-inducing. I would be happy to make another pass over it to improve the comments, and add some further test cases. A longer explanatory comment giving examples of expected results under various conditions might also be useful. Does this seem acceptable?

artwyman commented:
I'm convinced now that leaving the logic in PostGres is preferable.  I'm sorry I pushed things in the wrong direction.  I still don't trust myself to fully review the logic there.  It's primarily the timestamp handling, type conversion, and arithmetic I'm worried about, rather than the structure of the queries involved.  I don't know how to solve that discomfort.  You can work that out with a different reviewer, or decide it's time to go ahead and take the risk of it being imperfect.  It is security code, so the risk of attacks are what make me extra-nervous, but this is at least code which isn't directly subjected to user-controlled data, so it's not the worst risk.

I actually think putting the logic inside of a single query is preferable to a SQL function, because it's easier to change (no migration needed).  However I find that query to even harder to read and reason about, so your call.

I'm in favor of a proper token bucket with gradual refill.  I don't think the limited-time bucket approach is notably simpler.

One important change from my original review which I think is important: Safety of this algorithm only requires locking a single row, not the whole table.  That will address some of your deadlock/delay concerns.  I think the single-line INSERT + UPDATE you give above is implicitly based on a row lock.  In a larger transaction/function you can get one explicitly with SELECT FOR UPDATE.  Locking is a level of PostGres functionality I've done enough research to feel comfortable advising on.  I educated myself to be sure the revision handling was safe for sync.

A few points unrelated to this diff in particular, but which came up in your discussion:

- Locks are tied to a SQL connection, so it shouldn't be possible to deadlock in the case of a server crash, dropped connection, or transaction rollback  It certainly is possible for other code to run while the lock is held, which could indeed delay the release of the lock.  If we use a row lock, the scope of impact should be limited to a user+action pair.  We should also certainly avoid having independent code using the same sql connection in asynchronous ways, because SQL connections aren't thread-safe.

- Transactions are also tied to a connection.  I think that means our retry-on-connection-dropped is mostly useless, and potentially dangerous if it's applied individually to one request inside of a transaction.  Hopefully it's only being applied to cause retry of the transaction as a whole, which is safe, but I haven't reviewed all uses to be sure.

robknight commented:
> I actually think putting the logic inside of a single query is preferable to a SQL function, because it's easier to change (no migration needed). However I find that query to even harder to read and reason about, so your call.
> 
> I'm in favor of a proper token bucket with gradual refill. I don't think the limited-time bucket approach is notably simpler.

The latest version has a single query. It was eventually possible to come up with something that is both fairly short and fairly readable.

I've also added extensive comments and new test cases showing a) that this continues to work in all cases where the previous ones did, and b) showing some more detailed examples of how the refill mechanism behaves in practice.

artwyman commented on apps/passport-server/test/ratelimit.spec.ts:
```diff
@@ -1,18 +1,18 @@
 import { expect } from "chai";
 import "mocha";
 import { step } from "mocha-steps";
 import MockDate from "mockdate";
```
It looks like this file has several calls to MockDate.set() which don't have a corresponding call to MockDate.reset(), which I think can affect future tests.

robknight replied:
There's a reset in `afterAll()`, which contains it to the current test suite.

artwyman commented on apps/passport-server/test/ratelimit.spec.ts:
```diff
@@ -219,4 +268,122 @@ describe("generic rate-limiting features", function () {
+    };
+
+    for (let i = 1; i <= maxActions; i++) {
+      MockDate.set(Date.now() + 3600);
```
What's the intent of this addition?  I think Date.now() is in milliseconds, but 3600ms (3.6 seconds) is a weirdly arbitrary amount of time which suggests you might intend 1 hour (3600 seconds).  A rate of 10 actions per hour would allow for a new token every 6 minutes.

Separate but related, the docs don't make it clear to me whether MockData alters the behavior of Date.now() or only new Date().  Do you intend this to be adding to the existing mock date, or adding to the real current date?

robknight replied:
Replaced this with a more sensible constant.

I've also added a comment explaining the intended effect of `MockDate.set()` in these calls (and that it affects `Date.now()`).

artwyman commented on apps/passport-server/src/database/queries/rateLimit.ts:
```diff
@@ -1,40 +1,122 @@
+ * @param maxActions The maximum number of actions allowed in a time period
+ * @param startingActions The number of actions to start the bucket with
+ * @param timePeriod The time period in which maxActions are allowed
  * @returns Boolean indicating whether the action is permitted
```
Doesn't actually return a boolean.  Actually returns a bucket, requiring the caller to check the `remaining` value themselves.

robknight replied:
Fixed!

robknight commented on apps/passport-server/src/services/rateLimitService.ts:
```diff
@@ -91,10 +108,23 @@ export class RateLimitService {
+  /**
+   * Delete expired rate-limiting buckets from the DB.
+   */
+  public async pruneBuckets(): Promise<void> {
```
We could just write a migration for that.

robknight commented on apps/passport-server/src/database/queries/rateLimit.ts:
```diff
@@ -1,40 +1,122 @@
+  // 3) Return the bucket state
+  //
+  // Updates cover two fields:
+  // `remaining` needs to be updated to the number of tokens remaining *after*
```
I've added additional comments here.

robknight commented on apps/passport-server/src/database/queries/rateLimit.ts:
```diff
@@ -1,40 +1,122 @@
  * @param actionType The type of action being attempted
  * @param actionId The identifier of this action
+ * @param maxActions The maximum number of actions allowed in a time period
+ * @param startingActions The number of actions to start the bucket with
```
It's just more explicit than having `maxActions` be both the maximum and starting value. But perhaps it increases the surface area for mis-use?

robknight commented on apps/passport-server/src/database/models.ts:
```diff
@@ -339,3 +338,11 @@ export interface AnonMessageWithDetails extends AnonMessage {
+  action_id: string;
+  remaining: number;
+  // last_take is a bigint in Postgres, which node-postgres turns into a string
+  last_take: string;
```
I've made last_take non-null.

`passport-server` compiles in strict mode. `sqlQuery` returns something like `{ rows: any[] }`, which means that a row returned by `sqlQuery` will satisfy any type check, even in strict mode.

robknight commented on apps/passport-server/test/devconnect.spec.ts:
```diff
@@ -1725,9 +1714,10 @@ describe("devconnect functionality", function () {
 
-    await sleep(4000); // see env.ts for where this number comes from
+    // 5 resets are allowed per day, so the next one will be in one day / 5
+    MockDate.set(Date.now() + (86400 / 5) * 1000);
```
Added.

robknight commented on apps/passport-server/migrations/62_rate_limit_variable_periods.sql:
```diff
@@ -0,0 +1,15 @@
+DELETE FROM rate_limit_buckets;
+
+ALTER TABLE rate_limit_buckets DROP COLUMN last_refill;
+ALTER TABLE rate_limit_buckets ADD COLUMN last_take BIGINT;
```
It's now NOT NULL.

robknight commented on apps/passport-server/src/services/rateLimitService.ts:
```diff
@@ -25,8 +38,7 @@ export class RateLimitService {
     this.rollbarService = rollbarService;
     this.timeout = setTimeout(() => {
```
Done.

robknight commented on apps/passport-server/src/services/rateLimitService.ts:
```diff
@@ -1,20 +1,33 @@
   private timeout: NodeJS.Timeout;
   private readonly rollbarService: RollbarService | null;
   private disabled: boolean;
+  private buckets: Record<
```
Done

robknight commented on apps/passport-server/src/database/queries/rateLimit.ts:
```diff
@@ -1,40 +1,122 @@
+export async function pruneRateLimitBuckets(
+  client: Pool,
+  actionType: string,
+  expiredAgeMs: number
```
I've changed it to `expiryTimestampMs` and added a comment explaining that the caller should derive this value from the bucket configuration. If you know the bucket configuration, you know what the oldest possible bucket could be.

artwyman reviewed (approved):
I think the try/catch is the only comment I'm worried about at this point.
